A29-xds-tls-security.md

```diff
@@ -0,0 +1,411 @@
+
+If [match_subject_alt_names](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L373)
+is populated in the
+[combined_validation_context](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L251)
+of the received `UpstreamTlsContext` then gRPC validates the SAN entries in the server certificate
+by matching the `match_subject_alt_names` values using the
+[StringMatcher](https://github.com/envoyproxy/envoy/blob/main/api/envoy/type/matcher/string.proto#L20)
+semantics. This is called server authorization because this is how a client "authorizes" a server for
+the connection.
+
+The implementation is language dependent. As an example, Java uses an implementation
+of `X509TrustManager` through the `SslContext` provided to the client sub-channel. The other
+languages use similar hooks provided by the underlying TLS framework to implement server
+authorization.
+
+#### Server Side Flow
+
+The gRPC server xDS flow is described in
+[gRPC Server ](./A27-xds-global-load-balancing.md#grpc-client-architecture). The server
+configuration comes from the
+[`Listener`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/listener/v3/listener.proto#L39)
+resource which contains a list of
+[`FilterChain`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/listener/v3/listener_components.proto#L193)s.
+The `FilterChain` contains the `DownstreamTlsContext` that is applied to an incoming connection
+after selecting the best-matching `FilterChain` as described in
+[FilterChainMatch](A36-xds-for-servers.md#filterchainmatch). If the server is using `XdsServerCredentials`,
+gRPC processes the `DownstreamTlsContext` as described [above](#lds-for-server-side-security) and then
+uses the flow described above in
+[Use of Bootstrap File by the Certificate Provider Plugin Framework](#use-of-bootstrap-file-by-the-certificate-provider-plugin-framework)
+to fetch the requisite `CertificateProvider`s (one or two as the case may be). These `CertificateProvider`s,
+in one form or another, need to be made available to the incoming connection. Similar to the
+client side, this part of the flow is language dependent. For example, Java constructs a dynamic
+`SslContextProvider` that is directly usable by the connection's TLS handshaker to build an
+`SslContext` for the impending handshake.
+
+## Rationale
+
+gRPC has been steadily adding support for various xDS features including load balancing and
+advanced traffic management. xDS-based security was next in the roadmap and this proposal
+addresses that. gRPC makes a proxyless service mesh (PSM) possible. A secure proxyless
+service mesh requires the security features described in this proposal.
+
+The Certificate Provider Plugin Framework does not depend on an agent (SDS server). The
+SDS server/agent based solution compromised security (by sharing the private key with the
```

Comment:
There is still an agent here, in order to dump files on disk, so this seems questionable. The security property here is the same independent of whether it dumps files on disk or a UDS socket. I can understand if we say we want to _allow_ for systems that don't share the private key, although I think we believe the best security is for the workload to not have the private key at all and keep it solely within an agent.

Reply:
Yes, your comment applies to the file_watcher plugin but the "Certificate Provider Plugin Framework" itself does not assume the existence of an agent. For example, a plugin that mints CSRs and gets them signed by a CA. But I will modify the text to address your comment.

Reply:
fixed